src/vm.rs

```diff
@@ -517,6 +519,47 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
             .call_once(|| self.set_vcpu_htimedelta());
     }
 
+    // Completes any pending operation from the host for this VM.
+    fn complete_pending_op(&self, active_vcpu: &mut ActiveVmCpu<T>) -> EcallResult<u64> {
+        if let Some(pending_op) = active_vcpu.take_pending_op() {
```

Comment:
Ok I'll try that!
What's the clean way to retrieve the value of the A0 register with this approach? I was using the `active_vcpu` but with your proposal I won't have access to it when needed.

Reply:
IIUC, the thing you want to avoid is having A0 being read twice, right?

Could we pass a reference of `FinalizedVm` to `VmCpu::complete_pending_op()`? This way, we would invoke `complete_share_mem_region()` from inside VmCpu.
That being said, I'm not a big fan of this approach because `FinalizedVm` is the superset that contains `VmCpu` and I'm not sure it's clean to pass a reference down to `VmCpu`.

Another approach would be to call `VmCpu::complete_pending_op()` first, and expect it to return the value of the A0 register along with the `pending_op`, so that from `vm.rs`, we have all the information we need to call into `complete_share_mem_region()`. The only thing that is unclear to me with this approach is how we can make the return type of `VmCpu::complete_pending_op()` generic enough so that we wouldn't make it specific to the case we're trying to solve here.

Reply:
> IIUC, the thing you want to avoid is having A0 being read twice, right?

to be more precise, having the A0 that's being acted upon be different from the one that ends up getting written to the guest's register file. but yes, only reading it once would prevent that.

> That being said, I'm not a big fan of this approach because FinalizedVm is the superset that contains VmCpu and I'm not sure it's clean to pass a reference down to VmCpu.

agreed.

> Actually we can keep the take_pending_op() logic associated with calling VmCpu::complete_pending_op() first. That would simplify what needs to be returned by VmCpu::complete_pending_op() (still need to figure out how to make it generic enough).

i suppose that works too. another thing that comes to mind is using a closure that `complete_pending_op()` invokes on the ecall and the contents of A0. something like:

```
pub fn try_complete_pending_op<F, E>(&mut self, ecall_completion: F) -> Result<(), E>
where
    F: Fn(&SbiMessage, SbiReturnType) -> Result<(), E>
{
    match self.arch.pending_op {
        Some(Mmio(mmio_op)) => {
            // handle this same as before
        }
        Some(Ecall(msg)) => {
            let sbi_ret = ...;
            ecall_completion(&msg, sbi_ret)?;
            self.set_ecall_result(sbi_ret);
        }
        None => (),
    }
    self.arch.pending_op = None;
    Ok(())
}
```

and then call `try_complete_pending_op()` before entering the `active_vcpu.run()` loop in `run_vcpu()`.

Reply:
Okay I'll implement this way!